dot/sync/chain_sync_test.go

```diff
@@ -827,3 +827,189 @@ func TestChainSync_determineSyncPeers(t *testing.T) {
+					"idtest": {
+						who:    "idtest#1",
+						hash:   common.BytesToHash([]byte("hash#1")),
+						number: nil,
+					},
+				},
+			},
+			out: output{
+				highestBlock: int64(0),
+				err:          errNilBlockData,
+			},
+		},
+		{
+			name: "when has two peers (p1, p2) with p1.number 90 and p2.number 190 should return 190, nil",
+			in: input{
+				peerState: map[peer.ID]*peerState{
+					"idtest#1": {
+						who:    "idtest#1",
+						hash:   common.BytesToHash([]byte("hash#1")),
+						number: big.NewInt(90),
+					},
+					"idtest#2": {
+						who:    "idtest#2",
+						hash:   common.BytesToHash([]byte("hash#2")),
+						number: big.NewInt(190),
+					},
+				},
+			},
+			out: output{
+				highestBlock: int64(190),
+				err:          nil,
+			},
+		},
+		{
+			name: "when has two peers (p1, p2) with p1.number 190 and p2.number 90 should return 190, nil",
+			in: input{
+				peerState: map[peer.ID]*peerState{
+					"idtest#1": {
+						who:    "idtest#1",
+						hash:   common.BytesToHash([]byte("hash#1")),
+						number: big.NewInt(190),
+					},
+					"idtest#2": {
+						who:    "idtest#2",
+						hash:   common.BytesToHash([]byte("hash#2")),
+						number: big.NewInt(90),
+					},
+				},
+			},
+			out: output{
+				highestBlock: int64(190),
+				err:          nil,
+			},
+		},
+		{
+			name: "when has two peers (p1, p2) with p1.number nil and p2.number 90 should return 90, nil",
+			in: input{
+				peerState: map[peer.ID]*peerState{
+					"idtest#1": {
+						who:    "idtest#1",
+						hash:   common.BytesToHash([]byte("hash#1")),
+						number: nil,
+					},
+					"idtest#2": {
+						who:    "idtest#2",
+						hash:   common.BytesToHash([]byte("hash#2")),
+						number: big.NewInt(90),
+					},
+				},
+			},
+			out: output{
+				highestBlock: int64(90),
+				err:          nil,
+			},
+		},
+		{
+			name: "when has two peers (p1, p2) with p1.number 90 and p2.number nil should return 90, nil",
+			in: input{
+				peerState: map[peer.ID]*peerState{
+					"idtest#1": {
+						who:    "idtest#1",
+						hash:   common.BytesToHash([]byte("hash#1")),
+						number: big.NewInt(90),
+					},
+					"idtest#2": {
+						who:    "idtest#2",
+						hash:   common.BytesToHash([]byte("hash#2")),
+						number: nil,
+					},
+				},
+			},
+			out: output{
+				highestBlock: int64(90),
+				err:          nil,
+			},
+		},
+		{
+			name: "when has two peers (p1, p2) with p1.number nil and p2.number nil should return 0, errNilBlockData",
+			in: input{
+				peerState: map[peer.ID]*peerState{
+					"idtest#1": {
+						who:    "idtest#1",
+						hash:   common.BytesToHash([]byte("hash#1")),
+						number: nil,
+					},
+					"idtest#2": {
+						who:    "idtest#2",
+						hash:   common.BytesToHash([]byte("hash#2")),
+						number: nil,
+					},
+				},
+			},
+			out: output{
+				highestBlock: int64(0),
+				err:          errNilBlockData,
+			},
+		},
+	}
+
+	cs, _ := newTestChainSync(t) // just to improve tests performance
```

Comment:
How does this improve performance? If we gain just a few milliseconds, I would prefer to have this in the subtest so that subtests can be ran independently (no surprise) and in parallel (faster)

Reply:
I realized that is hard to measure it in my machine... Sometimes when I put it in the subtest is faster than out the subtest and vice versa... I will put it in the subtests.

Reply:
Done in d09bc0b